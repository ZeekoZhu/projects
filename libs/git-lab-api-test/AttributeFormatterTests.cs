using Projects.GitLabApi.Concrete;
using Snapshooter.NUnit;

namespace Projects.GitLabApiTest;

[TestFixture]
public class AttributeFormatterTests
{
  class RequestParams
  {
    public class HashParams
    {
      public string Key1 { get; set; }
      public string Key2 { get; set; }
    }

    public string SimpleString { get; set; }
    public string[] SimpleArray { get; set; }
    public DateTime SimpleDateTime { get; set; }
    public bool SimpleBool { get; set; }
    public int SimpleInt { get; set; }
    public long SimpleLong { get; set; }
    public double SimpleDouble { get; set; }
    public HashParams SimpleHash { get; set; }

    public List<HashParams> HashArray { get; set; }
  }

  [Test]
  public void CanFormatAttributes()
  {
    var formatter = new AttributeFormatter();
    var attributes = new RequestParams
    {
      SimpleString = "value1",
      SimpleArray = ["value1", "value2"],
      SimpleDateTime = new DateTime(2021, 1, 1),
      SimpleBool = true,
      SimpleInt = 1,
      SimpleLong = 1,
      SimpleDouble = 1,
      SimpleHash = new RequestParams.HashParams
      {
        Key1 = "value1",
        Key2 = "value2"
      },
      HashArray = new List<RequestParams.HashParams>
      {
        new()
        {
          Key1 = "value1",
          Key2 = "value2"
        },
        new()
        {
          Key1 = "value1",
          Key2 = "value2"
        }
      }
    };
    var formattedAttributes = formatter.FormatAttributes(attributes).ToList();

    Snapshot.Match(formattedAttributes);
  }

  class NullableTest
  {
    public int? NullableInt { get; set; }
    public string? NullableString { get; set; }
    public DateTime? NullableDateTime { get; set; }
    public bool? NullableBool { get; set; }
    public double? NullableDouble { get; set; }
    public RequestParams.HashParams? Params;
  }

  [Test]
  public void CanHandleNullableType()
  {
    var formatter = new AttributeFormatter();
    var attributes = new NullableTest
    {
      NullableInt = 1,
      NullableString = "value1",
      NullableDateTime = new DateTime(2021, 1, 1),
      NullableBool = true,
      NullableDouble = 1,
      Params = new RequestParams.HashParams
      {
        Key1 = "value1",
        Key2 = "value2"
      }
    };

    var formattedAttributes = formatter.FormatAttributes(attributes).ToList();

    Snapshot.Match(formattedAttributes);
  }

  [Test]
  public void ShouldIgnoreNullFields()
  {
    var formatter = new AttributeFormatter();
    var attributes = new NullableTest
    {
      NullableInt = 1,
      NullableString = null,
      NullableDateTime = new DateTime(2021, 1, 1),
      NullableBool = true,
      NullableDouble = 1,
      Params = null
    };

    var formattedAttributes = formatter.FormatAttributes(attributes).ToList();

    Snapshot.Match(formattedAttributes);
  }
}
