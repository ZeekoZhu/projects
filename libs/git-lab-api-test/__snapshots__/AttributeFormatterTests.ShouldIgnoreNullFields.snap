[
  {
    "Key": "nullable_int",
    "Value": "1"
  },
  {
    "Key": "nullable_date_time",
    "Value": "2021-01-01T00:00:00.000Z"
  },
  {
    "Key": "nullable_bool",
    "Value": "true"
  },
  {
    "Key": "nullable_double",
    "Value": "1"
  }
]
